File: src/vica_sensor_adapters/test/test_gyro_bias.py
"""정지 중 자이로 편향 추정기의 계약."""

import pytest

from vica_sensor_adapters.gyro_bias import GyroBiasEstimator


def _feed(estimator, count, value=(0.001, 0.002, 0.003)):
    for _ in range(count):
        estimator.add(*value)
    return estimator


def test_before_enough_samples_it_does_not_correct():
    """표본이 모자라면 보정하지 않는다. 원값을 그대로 돌려준다."""
    est = _feed(GyroBiasEstimator(sample_count=10, max_abs_rate=0.05), 9)

    assert not est.ready
    assert est.correct(0.5, 0.6, 0.7) == (0.5, 0.6, 0.7)


def test_exactly_enough_samples_becomes_ready():
    """경계값: 표본이 정확히 sample_count면 준비된다."""
    est = _feed(GyroBiasEstimator(sample_count=10, max_abs_rate=0.05), 10)

    assert est.ready


def test_bias_is_the_mean_of_collected_samples():
    """편향은 수집 구간의 축별 평균이다."""
    est = GyroBiasEstimator(sample_count=4, max_abs_rate=0.05)
    for gz in (0.001, 0.002, 0.003, 0.004):
        est.add(0.0, 0.0, gz)

    bx, by, bz = est.bias
    assert bx == pytest.approx(0.0)
    assert by == pytest.approx(0.0)
    assert bz == pytest.approx(0.0025)


def test_correction_subtracts_the_bias():
    """준비된 뒤에는 원값에서 편향을 뺀다."""
    est = _feed(GyroBiasEstimator(sample_count=5, max_abs_rate=0.05), 5,
                value=(0.0, 0.0, 0.001))

    _, _, gz = est.correct(0.010, 0.020, 0.030)
    assert gz == pytest.approx(0.029)


def test_motion_during_calibration_aborts_permanently():
    """보정 중 움직이면 포기한다. 틀린 편향을 박아넣지 않는다."""
    est = GyroBiasEstimator(sample_count=10, max_abs_rate=0.05)
    _feed(est, 5)
    est.add(0.0, 0.0, 0.5)
    _feed(est, 100)

    assert est.aborted
    assert not est.ready
    assert est.correct(0.5, 0.6, 0.7) == (0.5, 0.6, 0.7)


def test_rate_exactly_at_threshold_is_accepted():
    """경계값: 임계값과 같은 각속도는 정지로 본다."""
    est = GyroBiasEstimator(sample_count=2, max_abs_rate=0.05)
    est.add(0.0, 0.0, 0.05)
    est.add(0.0, 0.0, 0.05)

    assert not est.aborted
    assert est.ready


def test_rate_just_over_threshold_aborts():
    """경계값: 임계값을 조금이라도 넘으면 포기한다."""
    est = GyroBiasEstimator(sample_count=2, max_abs_rate=0.05)
    est.add(0.0, 0.0, 0.0500001)

    assert est.aborted


def test_threshold_applies_to_every_axis():
    """임계 판정은 세 축 각각에 적용한다."""
    est = GyroBiasEstimator(sample_count=2, max_abs_rate=0.05)
    est.add(0.9, 0.0, 0.0)

    assert est.aborted


def test_zero_sample_count_disables_the_feature():
    """sample_count가 0이면 기능을 끈다. 보정도 포기도 하지 않는다."""
    est = _feed(GyroBiasEstimator(sample_count=0, max_abs_rate=0.05), 50)

    assert not est.ready
    assert not est.aborted
    assert est.correct(0.5, 0.6, 0.7) == (0.5, 0.6, 0.7)


def test_extra_samples_after_ready_do_not_change_bias():
    """준비된 뒤 들어오는 값은 편향을 바꾸지 않는다. 주행 중 값이 섞이면 안 된다."""
    est = _feed(GyroBiasEstimator(sample_count=3, max_abs_rate=0.05), 3,
                value=(0.0, 0.0, 0.001))
    before = est.bias

    est.add(0.0, 0.0, 0.04)
    assert est.bias == before


def test_progress_reports_collected_and_target():
    """진행 상황을 읽을 수 있어야 로그와 진단에 남길 수 있다."""
    est = _feed(GyroBiasEstimator(sample_count=10, max_abs_rate=0.05), 4)

    assert est.collected == 4
    assert est.sample_count == 10


def _quiet(estimator, count, value=0.01):
    """정차로 인정될 만큼 조용한 표본."""
    for _ in range(count):
        estimator.add(value, value, value)


def _shaky(estimator, count, center=0.01, amp=0.03):
    """직진 중처럼 흔들리는 표본. 평균은 center 지만 폭이 크다."""
    for i in range(count):
        d = amp if i % 2 == 0 else -amp
        estimator.add(center + d, center + d, center + d)


def _ready_estimator(bias=0.01, **kw):
    """초기 확정을 마친 추정기를 만든다."""
    est = GyroBiasEstimator(sample_count=10, max_abs_rate=0.05, **kw)
    _feed(est, 10, (bias, bias, bias))
    assert est.ready
    return est


def test_refresh_off_by_default_keeps_old_behaviour():
    """refresh_sample_count 0 이면 종전과 똑같이 한 번 확정하고 고정이다."""
    est = _ready_estimator(bias=0.01, refresh_sample_count=0)
    before = est.bias
    _quiet(est, 500, value=0.02)
    assert est.bias == before


def test_quiet_stretch_updates_the_bias():
    """정차가 충분히 이어지면 편향을 갱신한다."""
    est = _ready_estimator(bias=0.01, refresh_sample_count=20,
                           refresh_alpha=0.5, max_abs_dev=0.01)
    _quiet(est, 20, value=0.02)
    assert est.bias[2] == pytest.approx(0.015, abs=1e-9)


def test_update_is_blended_not_replaced():
    """한 번의 측정이 편향을 통째로 갈아치우지 않는다."""
    est = _ready_estimator(bias=0.0, refresh_sample_count=10,
                           refresh_alpha=0.2, max_abs_dev=0.01,
                           max_refresh_jump=1.0)
    _quiet(est, 10, value=0.04)
    assert est.bias[2] == pytest.approx(0.008, abs=1e-9)


def test_shaky_stretch_is_rejected():
    """직진처럼 흔들리는 구간은 정차로 보지 않는다 — 이 시험이 핵심이다."""
    est = _ready_estimator(bias=0.01, refresh_sample_count=20,
                           refresh_alpha=0.5, max_abs_dev=0.01)
    before = est.bias
    _shaky(est, 200, center=0.05, amp=0.03)
    assert est.bias == before


def test_motion_resets_the_stretch():
    """중간에 움직이면 모으던 표본을 버린다. 이어붙이지 않는다."""
    est = _ready_estimator(bias=0.01, refresh_sample_count=20,
                           refresh_alpha=0.5, max_abs_dev=0.01)
    before = est.bias
    _quiet(est, 15, value=0.02)
    est.add(0.5, 0.0, 0.0)
    _quiet(est, 15, value=0.02)
    assert est.bias == before


def test_sudden_jump_is_refused():
    """옛 값과 너무 다르면 그 회차를 버린다."""
    est = _ready_estimator(bias=0.01, refresh_sample_count=10,
                           refresh_alpha=0.5, max_abs_dev=0.01,
                           max_refresh_jump=0.02)
    before = est.bias
    _quiet(est, 10, value=0.04)
    assert est.bias == before


def test_refresh_works_even_if_startup_aborted():
    """기동 때 못 쟀어도 첫 정차에서 잡는다."""
    est = GyroBiasEstimator(sample_count=10, max_abs_rate=0.05,
                            refresh_sample_count=10, refresh_alpha=1.0,
                            max_abs_dev=0.01)
    est.add(0.5, 0.0, 0.0)
    assert est.aborted and not est.ready
    _quiet(est, 10, value=0.02)
    assert est.ready
    assert est.bias[2] == pytest.approx(0.02, abs=1e-9)


def test_refresh_count_is_reported():
    """몇 번 갱신했는지 볼 수 있어야 실주행에서 동작을 확인한다."""
    est = _ready_estimator(bias=0.01, refresh_sample_count=10,
                           refresh_alpha=0.5, max_abs_dev=0.01)
    assert est.refresh_count == 0
    _quiet(est, 10, value=0.02)
    assert est.refresh_count == 1

File: src/vica_sensor_adapters_cpp/test/test_gyro_bias.cpp
#include <gtest/gtest.h>

#include "vica_sensor_adapters_cpp/gyro_bias.hpp"

using vica_sensor_adapters_cpp::GyroBiasEstimator;
using vica_sensor_adapters_cpp::Vec3;

namespace
{

void feed(GyroBiasEstimator & est, int count, double x = 0.001, double y = 0.002,
  double z = 0.003)
{
  for (int i = 0; i < count; ++i) {
    est.add(x, y, z);
  }
}

/// 정차로 인정될 만큼 조용한 표본.
void quiet(GyroBiasEstimator & est, int count, double value = 0.01)
{
  for (int i = 0; i < count; ++i) {
    est.add(value, value, value);
  }
}

/// 직진 중처럼 흔들리는 표본. 평균은 center 지만 폭이 크다.
void shaky(GyroBiasEstimator & est, int count, double center = 0.01, double amp = 0.03)
{
  for (int i = 0; i < count; ++i) {
    const double d = (i % 2 == 0) ? amp : -amp;
    est.add(center + d, center + d, center + d);
  }
}

/// 초기 확정을 마친 추정기.
GyroBiasEstimator ready_estimator(
  double bias = 0.01, int refresh_sample_count = 0, double refresh_alpha = 0.2,
  double max_abs_dev = 0.01, double max_refresh_jump = 0.02)
{
  GyroBiasEstimator est(10, 0.05, refresh_sample_count, refresh_alpha,
    max_abs_dev, max_refresh_jump);
  feed(est, 10, bias, bias, bias);
  EXPECT_TRUE(est.ready());
  return est;
}

}

TEST(GyroBias, BeforeEnoughSamplesItDoesNotCorrect)
{
  GyroBiasEstimator est(10, 0.05);
  feed(est, 9);

  EXPECT_FALSE(est.ready());
  const Vec3 out = est.correct(0.5, 0.6, 0.7);
  EXPECT_DOUBLE_EQ(out[0], 0.5);
  EXPECT_DOUBLE_EQ(out[1], 0.6);
  EXPECT_DOUBLE_EQ(out[2], 0.7);
}

TEST(GyroBias, ExactlyEnoughSamplesBecomesReady)
{
  GyroBiasEstimator est(10, 0.05);
  feed(est, 10);

  EXPECT_TRUE(est.ready());
}

TEST(GyroBias, BiasIsTheMeanOfCollectedSamples)
{
  GyroBiasEstimator est(4, 0.05);
  for (double gz : {0.001, 0.002, 0.003, 0.004}) {
    est.add(0.0, 0.0, gz);
  }

  const Vec3 b = est.bias();
  EXPECT_NEAR(b[0], 0.0, 1e-12);
  EXPECT_NEAR(b[1], 0.0, 1e-12);
  EXPECT_NEAR(b[2], 0.0025, 1e-12);
}

TEST(GyroBias, CorrectionSubtractsTheBias)
{
  GyroBiasEstimator est(5, 0.05);
  feed(est, 5, 0.0, 0.0, 0.001);

  const Vec3 out = est.correct(0.010, 0.020, 0.030);
  EXPECT_NEAR(out[2], 0.029, 1e-12);
}

TEST(GyroBias, MotionDuringCalibrationAbortsPermanently)
{
  GyroBiasEstimator est(10, 0.05);
  feed(est, 5);
  est.add(0.0, 0.0, 0.5);
  feed(est, 100);

  EXPECT_TRUE(est.aborted());
  EXPECT_FALSE(est.ready());
  const Vec3 out = est.correct(0.5, 0.6, 0.7);
  EXPECT_DOUBLE_EQ(out[0], 0.5);
  EXPECT_DOUBLE_EQ(out[2], 0.7);
}

TEST(GyroBias, RateExactlyAtThresholdIsAccepted)
{
  GyroBiasEstimator est(2, 0.05);
  est.add(0.0, 0.0, 0.05);
  est.add(0.0, 0.0, 0.05);

  EXPECT_FALSE(est.aborted());
  EXPECT_TRUE(est.ready());
}

TEST(GyroBias, RateJustOverThresholdAborts)
{
  GyroBiasEstimator est(2, 0.05);
  est.add(0.0, 0.0, 0.0500001);

  EXPECT_TRUE(est.aborted());
}

TEST(GyroBias, ThresholdAppliesToEveryAxis)
{
  GyroBiasEstimator est(2, 0.05);
  est.add(0.9, 0.0, 0.0);

  EXPECT_TRUE(est.aborted());
}

TEST(GyroBias, ZeroSampleCountDisablesTheFeature)
{
  GyroBiasEstimator est(0, 0.05);
  feed(est, 50);

  EXPECT_FALSE(est.ready());
  EXPECT_FALSE(est.aborted());
  const Vec3 out = est.correct(0.5, 0.6, 0.7);
  EXPECT_DOUBLE_EQ(out[1], 0.6);
}

TEST(GyroBias, ExtraSamplesAfterReadyDoNotChangeBias)
{
  GyroBiasEstimator est(3, 0.05);
  feed(est, 3, 0.0, 0.0, 0.001);
  const Vec3 before = est.bias();

  est.add(0.0, 0.0, 0.04);
  EXPECT_DOUBLE_EQ(est.bias()[2], before[2]);
}

TEST(GyroBias, ProgressReportsCollectedAndTarget)
{
  GyroBiasEstimator est(10, 0.05);
  feed(est, 4);

  EXPECT_EQ(est.collected(), 4);
  EXPECT_EQ(est.sample_count(), 10);
}

TEST(GyroBias, RefreshOffByDefaultKeepsOldBehaviour)
{
  GyroBiasEstimator est = ready_estimator(0.01, 0);
  const Vec3 before = est.bias();
  quiet(est, 500, 0.02);

  EXPECT_DOUBLE_EQ(est.bias()[2], before[2]);
}

TEST(GyroBias, QuietStretchUpdatesTheBias)
{
  GyroBiasEstimator est = ready_estimator(0.01, 20, 0.5, 0.01);
  quiet(est, 20, 0.02);

  EXPECT_NEAR(est.bias()[2], 0.015, 1e-9);
}

TEST(GyroBias, UpdateIsBlendedNotReplaced)
{
  GyroBiasEstimator est = ready_estimator(0.0, 10, 0.2, 0.01, 1.0);
  quiet(est, 10, 0.04);

  EXPECT_NEAR(est.bias()[2], 0.008, 1e-9);
}

TEST(GyroBias, ShakyStretchIsRejected)
{
  GyroBiasEstimator est = ready_estimator(0.01, 20, 0.5, 0.01);
  const Vec3 before = est.bias();
  shaky(est, 200, 0.05, 0.03);

  EXPECT_DOUBLE_EQ(est.bias()[2], before[2]);
}

TEST(GyroBias, MotionResetsTheStretch)
{
  GyroBiasEstimator est = ready_estimator(0.01, 20, 0.5, 0.01);
  const Vec3 before = est.bias();
  quiet(est, 15, 0.02);
  est.add(0.5, 0.0, 0.0);
  quiet(est, 15, 0.02);

  EXPECT_DOUBLE_EQ(est.bias()[2], before[2]);
}

TEST(GyroBias, SuddenJumpIsRefused)
{
  GyroBiasEstimator est = ready_estimator(0.01, 10, 0.5, 0.01, 0.02);
  const Vec3 before = est.bias();
  quiet(est, 10, 0.04);

  EXPECT_DOUBLE_EQ(est.bias()[2], before[2]);
}

TEST(GyroBias, RefreshWorksEvenIfStartupAborted)
{
  GyroBiasEstimator est(10, 0.05, 10, 1.0, 0.01);
  est.add(0.5, 0.0, 0.0);
  ASSERT_TRUE(est.aborted());
  ASSERT_FALSE(est.ready());

  quiet(est, 10, 0.02);
  EXPECT_TRUE(est.ready());
  EXPECT_NEAR(est.bias()[2], 0.02, 1e-9);
}

TEST(GyroBias, RefreshCountIsReported)
{
  GyroBiasEstimator est = ready_estimator(0.01, 10, 0.5, 0.01);
  EXPECT_EQ(est.refresh_count(), 0);

  quiet(est, 10, 0.02);
  EXPECT_EQ(est.refresh_count(), 1);
}
